src/components/OnboardingManager.tsx

```diff
@@ -1,114 +1,97 @@
-      f => `/onboarding/${f.url}` === pathname,
-    );
-    return find;
-  }, [pathname, onboarding]);
-
-  useEffect(() => {
-    if (!onboarding) return;
+  const auth = useAuth();
+  const profile = useProfile();
+  const onboarding = useGuardedOnboarding();
 
-    // out of bounds redirect
-    if (
-      onboarding.progress < 0 ||
-      onboarding.progress >= onboarding.flow.length ||
-      !auth?.userId ||
-      (profile?.profileReady && profile?.profileData)
-    ) {
-      router.push(CONFIG.homepage);
-      return;
-    }
+  const { pageProgress } = useOnboardingNavigation();
+  const { push } = useRouter();
 
-    if (pageProgress > onboarding.progress)
-      router.push(`/onboarding/${onboarding.flow[onboarding.progress].url}`);
-  }, [onboarding, profile, router, pageProgress]);
+  if (!profile)
+    throw new Error(
+      'Fatal: onboarding manager should be wrapped inside profile context',
+    );
 
-  const advanceProgress = () => {
-    // safeguard
-    if (!onboarding) return;
-    if (pageProgress > onboarding.progress) return;
+  if (!auth)
+    throw new Error(
+      'Fatal: onboarding manager should be wrapped inside auth context',
+    );
 
-    const newProgress = pageProgress + 1;
+  const {
+    flow: onboardingFlow,
+    progress: onboardingProgress,
+    setProgress,
+  } = onboarding;
+  const { profileReady, profileData } = profile;
+  const { userId } = auth;
 
+  useEffect(() => {
     if (
-      newProgress > onboarding.progress &&
-      onboarding.progress < onboarding.flow.length - 1
+      // edge case (realistically shouldn't happen)
+      onboardingProgress < 0 ||
+      // progress beyond length of flow
+      (onboardingFlow.length !== 0 &&
+        onboardingProgress >= onboardingFlow.length) ||
+      // already onboarded or not signed in
+      (profileReady && (profileData || !userId))
     ) {
-      onboarding.setCanContinue(false);
-      onboarding.setProgress(newProgress);
+      push(CONFIG.homepage);
+      return;
     }
 
-    if (newProgress >= onboarding.flow.length) {
-      onboarding.flushData().then(() => {
-        router.push(CONFIG.homepage);
-      });
-    } else {
-      router.push(`/onboarding/${onboarding.flow[newProgress].url}`);
-    }
+    // advance progress if exactly 1 more
+    if (pageProgress === onboardingProgress + 1) setProgress(pageProgress);
+    // otherwise, rebound to current progress
+    else if (pageProgress > onboardingProgress)
+      push(`/onboarding/${onboardingFlow[onboardingProgress].url}`);
+  }, [
+    onboardingProgress,
+    onboardingFlow,
+    profileReady,
+    profileData,
+    push,
+    pageProgress,
+    userId,
+    setProgress,
+  ]);
+
+  const goToHomepage = () => {
+    // TODO: add form dirty bool to context to conditionally show dialog
+    // eslint-disable-next-line no-alert
+    const ans = window.confirm(
+      'Are you sure you want to continue? Your changes will not be saved after you leave.',
```

Comment:
Is there a way to edit out the "localhost:300 says" part?

<img width="466" alt="image" src="https://github.com/calblueprint/immigration-justice-project/assets/84427964/5a40a5ec-ef0c-4a2f-9bbd-0aae285881e5">

Reply:
I don't think it's possible, since the "unchangeableness" of it indicates that it must be legitimate, as it is generated by the browser (cannot be faked by UI) ([related](https://stackoverflow.com/a/43959/22063638))

That being said, we could make it a UI modal, since the effects of this doesn't really depend on the browser. It shouldn't be too difficult if we can use Radix primitives to abstract away the functionalities for us. We may need to consult @kyrenetam for a design though.

Reply:
ok, we can keep it like this for now since designs might take a while if we're able to get one.